sklearn/utils/set_output.py

```diff
@@ -0,0 +1,225 @@
+    """Get output configure based on estimator and global configuration.
+
+    Parameters
+    ----------
+    estimator : estimator instance, default=None
+        If not `None`, check the estimator for output container.
+
+    method : str
+        Method to get container output for.
+
+    Returns
+    -------
+    config : Bunch
+        Dictionary with keys, "dense", that specifies the container for `method`.
+    """
+    est_sklearn_output_config = getattr(estimator, "_sklearn_output_config", {})
+    if method in est_sklearn_output_config:
+        container_str = est_sklearn_output_config[method]
+    else:
+        container_str = get_config()[f"output_{method}"]
+
+    return Bunch(**{"dense": container_str})
+
+
+def _wrap_output_with_container(
+    estimator, output, method, index, constructor_kwargs=None
+):
+    """Wrap output with container based on an estimator's or global config.
+
+    Parameters
+    ----------
+    estimator : estimator instance
+        Estimator to get the output configuration from.
+
+    output : ndarray
+        Output to to wrap with container.
+
+    method : str
+        Method that returned `output`.
+
+    index : array-like
+        Index to attach to output.
+
+    constructor_kwargs : dict, default=None
+        Keyword arguments passed to container constructor.
+
+    Returns
+    -------
+    wrapped_output : ndarray or DataFrame
+        Wrapped output with column names and index or `output` itself if wrapping is
+        not configured.
+    """
+    output_container = get_output_config(estimator, method)
+    return make_named_container(
+        output=output,
+        index=index,
+        columns=getattr(estimator, "get_feature_names_out", None),
+        dense_container=output_container.dense,
+        constructor_kwargs=constructor_kwargs,
+    )
+
+
+def _wrap_method_output(f, method):
+    """Wrapper used by SetOutputMixin to automatically wrap methods."""
+
+    @wraps(f)
+    def wrapped(self, X, *args, **kwargs):
+        output = f(self, X, *args, **kwargs)
+        return _wrap_output_with_container(
+            self, output, method, getattr(X, "index", None)
+        )
+
+    return wrapped
+
+
+class SetOutputMixin:
+    """Mixin that dynamically wraps methods to return container based on config.
+
+    Currently `SetOutputMixin` wraps `transform` and `fit_transform` and configures
+    it based on `set_output` of the global configuration.
+    """
+
+    def __init_subclass__(cls, **kwargs):
+        # Dynamically wraps `transform` and `fit_transform` and configure it's
+        # output based on `set_output`.
+        if hasattr(cls, "transform"):
+            cls.transform = _wrap_method_output(cls.transform, "transform")
+        if hasattr(cls, "fit_transform"):
+            cls.fit_transform = _wrap_method_output(cls.fit_transform, "transform")
+
+    def set_output(self, *, transform=None):
+        """Set output container.
+
```

Comment:
I think we should mark this method experimental to let the user known that the API might change without going through the usual deprecation cycle at this point.

Reply:
This method been marked as experimental with a `..note::`